### PR TITLE
✨ Backfills

### PR DESCRIPTION
**Pull request by RTLS**

### CODE REVIEW

**Comment by RTLS** [comment]
* **#634** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/634?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/634?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: lib/sequin/consumers/backfill.ex [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+      :sink_consumer_id,
+      :state,
+      :rows_initial_count,
+      :initial_min_cursor
+    ])
+    |> validate_required([:account_id, :sink_consumer_id, :state, :initial_min_cursor])
+    |> foreign_key_constraint(:sink_consumer_id)
+    |> unique_constraint(:sink_consumer_id,
+      name: "backfills_sink_consumer_id_index",
+      message: "already has an active backfill"
```
Comment (acco) [inline lib/sequin/consumers/backfill.ex:58]: nice

File: lib/sequin/consumers/backfill.ex [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+    field :state, Ecto.Enum,
+      values: [:active, :completed, :cancelled],
+      default: :active
+
+    field :rows_initial_count, :integer
+    field :rows_processed_count, :integer, default: 0
+    field :rows_ingested_count, :integer, default: 0
+    field :completed_at, :utc_datetime_usec, read_after_writes: true
+    field :canceled_at, :utc_datetime_usec, read_after_writes: true
```
Comment (acco) [inline lib/sequin/consumers/backfill.ex:40]: good data model 👍

File: lib/sequin/consumers/sink_consumer.ex [hunk 3]
```diff
@@ -54,6 +54,7 @@ defmodule Sequin.Consumers.SinkConsumer do
 
     embeds_many :source_tables, SourceTable, on_replace: :delete
     embeds_one :record_consumer_state, RecordConsumerState, on_replace: :delete
+    has_one :active_backfill, Sequin.Consumers.Backfill, where: [state: :active]
```
Comment (acco) [inline lib/sequin/consumers/sink_consumer.ex:57]: nice

File: priv/repo/migrations/20241120224011_create_backfills_table.exs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+    RETURNS TRIGGER AS $$
+    BEGIN
+      IF NEW.state = 'completed' AND OLD.state != 'completed' THEN
+        NEW.completed_at = NOW();
+      ELSIF NEW.state = 'cancelled' AND OLD.state != 'cancelled' THEN
+        NEW.canceled_at = NOW();
+      END IF;
+      RETURN NEW;
+    END;
+    $$ LANGUAGE plpgsql;
```
Comment (acco) [inline priv/repo/migrations/20241120224011_create_backfills_table.exs:56]: nice

**Review by acco — APPROVED** [review]

**Pull request merged**